srcbpatch/streamreplacer.cpp

```diff
@@ -0,0 +1,544 @@
+    return unique_ptr<StreamReplacer>(new UsualReplacer(src, trg));
+}
+
+
+//--------------------------------------------------
+///
+///      |--SRC 1  TRG 1  |
+///  O - |-- ...          | - o
+///      |--SRC N  TRG N  |
+/// 
+class ChoiceReplacer final : public ReplacerWithNext
+{
+    typedef struct
+    {
+        span<const char> src_;
+        span<const char> trg_;
+    }ChoiceReplacerPair;
+
+public:
+    /// <summary>
+    ///   creating ChoiceReplacer from provided pairs
+    /// </summary>
+    /// <param name="choice">vector os source & target pairs</param>
+    ChoiceReplacer(StreamReplacerChoice& choice)
+    {
+        size_t bufferSize = 0; // to allocate buffer
+        const size_t sz = choice.size();
+        rpairs_.resize(sz);
+        for (size_t i = 0; i < sz; ++i)
+        {
+            auto& vPair = choice[i]; // copy from
+            auto& rpair = rpairs_[i];// copy to
+
+            rpair.src_ = vPair.first->access();
+            const size_t sourceSize = rpair.src_.size();
+            if (bufferSize < sourceSize)
+            {
+                bufferSize = sourceSize; // calculate necessary buffer size
+            }
+
+            rpair.trg_ = vPair.second->access();
+        }
+
+        buffer_.resize(bufferSize + 1); // +1 needed for inner logic but never being accessed
+    }
+
+    void DoReplacements(const char toProcess, const bool aEod) const override;
+
+protected:
+    /// <summary>
+    ///    check for partial or full match of data from buffer_
+    ///       with any of lexemes sequentially
+    ///       and provides type of match and matched length and index if found
+    /// </summary>
+    /// <param name="indexFrom"> search in pairs from this index</param>
+    /// <returns>partial, full, length, index </returns>
+    tuple<bool, bool, size_t, size_t> CheckMatch(const size_t indexFrom) const
+    {
+        size_t i = indexFrom;
+        for (; i < rpairs_.size(); ++i)
+        {
+            const auto& srcSpan = rpairs_[i].src_;
+            const size_t cmpLength = (srcSpan.size() > cachedAmount_) ? cachedAmount_ : srcSpan.size();
+
+            if (0 != memcmp(srcSpan.data(), buffer_.data(), cmpLength))
+            {
+                continue; // no match here
+            }
+
+            const bool fullMatch = cmpLength == srcSpan.size();
+            return {!fullMatch, fullMatch, cmpLength, i};
+        }
+        return {false, false, 0, 0};
+    }
+
+protected:
+    // our pairs sorted by priority - only one of them could be replaced for concrete pos
+    vector<ChoiceReplacerPair> rpairs_;
+
+    mutable size_t cachedAmount_ = 0; // we cache this amount of data
+    mutable size_t indexOfCached_ = 0; // at this index from rpairs_
+
+    // this is used to hold temporary data while the logic is 
+    // looking for the new beginning of the cached value
+    mutable vector<char> buffer_;
+};
+
+
+void ChoiceReplacer::DoReplacements(const char toProcess, const bool aEod) const
+{
```

Comment:
Well - yes. logic has a lot of logical branches.
You are actually, knows what it is doing.

can you reproduce the same - but shorter?
PS. I have rewrite this about 5 times. every time after creation of new unit test. For you it can be a great practice. ;)

Reply:
I don't think I can rewrite it shorter, but what I would do - is to implement a couple of inline functions to make this method easy to understand. The method will be shorter, but in total it will be larger. 

PS I know that is going on here, but I had to read this method several times to understand it, and even now I need to read it again to remember. It's quiet big, you know. And the code isn't simple. Adding inline functions would greatly enhance understanding.

Reply:
Example of pseudocode

```
void ChoiceReplacer::DoReplacements(const char toProcess, const bool aEod) const{
    if (nullptr == pNext_)
    {
        throw logic_error("Replacement chain has been broken. Communicate with maintainer");
    }
    if (aEod)
        ProcessLastCharacter(toProcess);
    else
        ProcessCharacter(toProcess);
}

void ChoiceReplacer::ProcessLastCharacter(const char toProcess) const{
        while (cachedAmount_ > 0)
        {
            auto [IsFullMatch,  SrcMatchedLength, DestString] = FinfMatchInReplacaments();
            if (IsFullMatch){
                    MakeReplace(DestString); // MakeReplace accepts std::string_view, or char
                    CleanTheCache(SrcMatchedLength);
            }
            else{
                    MakeReplace(buffer_.front();
                    CleanTheCache(1);
            }
        };

        pNext_->DoReplacements(toProcess, aEod); // send end of the data further
        return;
}

void ChoiceReplacer::ProcessCharacter(const char toProcess) const{
    buffer_.data()[cachedAmount_++] = toProcess;
    while (cachedAmount_ > 0)
    {
            auto [IsFullMatch,  SrcMatchedLength, DestString] = FinfMatchInReplacaments();
            if (IsFullMatch){
                    MakeReplace(DestString); // MakeReplace accepts std::string_view, or char
                    CleanTheCache(SrcMatchedLength);
            }
            else{
                    SavePartialReplace(index);
            }
        }
        MakeReplace(buffer_.front();
        CleanTheCache(1);
    } 
}
```